rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,968 @@
+use crate::cells::bandwidth::queue::PacketQueue;
+use crate::cells::{Cell, Packet};
+use crate::core::CALIBRATED_START_INSTANT;
+use crate::error::Error;
+use crate::metal::timer::Timer;
+use async_trait::async_trait;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use std::cmp::{max, min};
+use std::fmt::Debug;
+use std::sync::atomic::AtomicI32;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio::time::{Duration, Instant};
+use tracing::debug;
+
+use super::bandwidth::queue::{DropTailQueue, DropTailQueueConfig};
+use super::bandwidth::LARGE_DURATION;
+use super::{ControlInterface, Egress, Ingress};
+
+pub const LARGE_RATE_LIMIT: u64 = 1073741824;
+pub const LARGE_BUFFER: usize = 1073741824;
+
+// transfer byte into time
+fn length_to_time(length: usize, rate: u64) -> Duration {
+    if length == 0 {
+        Duration::from_secs(0)
+    } else if rate == 0 {
+        LARGE_DURATION
+    } else {
+        Duration::from_secs_f64(length as f64 / rate as f64)
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Debug, Clone, Copy)]
+pub struct TokenBucket {
+    buffer: usize,
+    rate: u64,
+    tokens: i64,
```

Comment:
Does the negative value have any meaning? If yes, then why we not use Option for this in Rust? If not, why not u64?

Also, why we cannot directly store tokens in type `Duration`?

Reply:
`tokens` here does not need to be negative. But I need to use negative values in `dequeue`, so I set `tokens` here to be `i64`.